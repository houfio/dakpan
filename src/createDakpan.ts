import {
  Component,
  ComponentClass,
  ConsumerProps,
  createContext,
  ProviderProps,
  ReactElement,
  ReactNodeArray,
  ReactPortal,
  StatelessComponent
} from 'react';

import { createConsumer } from './createConsumer';
import { createProvider } from './createProvider';
import { Actions, DakpanProviderProps, GetState, MappedActions, SetState } from './types';
import { withDakpan } from './withDakpan';

export const createDakpan = <S>(initialState: S) => <A extends Actions<S>>(actions: A) => {
  let getState: GetState<S> | undefined;
  let setState: SetState<S, DakpanProviderProps> | undefined;

  const mergeAction = (result: Partial<S>) => {
    if (!getState || !setState) {
      throw new Error('Provider unmounted before action could complete');
    } else if (typeof result !== 'object') {
      throw new Error('Actions may only return objects');
    }

    const newState = {
      ...getState() as any,
      ...result as any
    };

    setState(newState);

    return newState;
  };

  const { Provider, Consumer } = createContext<S>(initialState);
  const mappedActions = Object.keys(actions).reduce(
    (result, key) => {
      const e = (...args: any[]) => () => {
        const action = actions[key];

        if (!getState || !setState) {
          throw new Error('You may not dispatch an action before its provider is mounted');
        }

        const state = getState();
        const result = action(...args)(state);

        if (result instanceof Promise) {
          return result.then(mergeAction);
        }

        return mergeAction(result);
      };

      const action: any = (...args: any[]) => e(...args)();
      action.e = e;

      return {
        ...result,
        [key]: action
      };
    },
    {}
  ) as MappedActions<S, A>;

  const provider = createProvider(Provider, initialState, (get, set) => {
    getState = get;
    setState = set;
  });
  const consumer = createConsumer(Consumer, mappedActions);

  return {
    Provider: provider,
    Consumer: consumer,
    actions: mappedActions,
    withDakpan: withDakpan(consumer)
  };
};
