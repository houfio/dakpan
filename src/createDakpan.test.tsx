import { mount } from 'enzyme';
import * as React from 'react';
import { Component } from 'react';

import { createDakpan } from './createDakpan';

const createMockDakpan = () => {
  const mockUpdate = jest.fn((current: number) => new Promise<number>((resolve) => resolve(current + 1)));
  const mockAction = jest.fn();

  return {
    ...createDakpan({
      hello: 'world',
      test: 0
    })({
      append: (append: string) => ({ hello }) => ({
        hello: `${hello}${append}`
      }),
      increment: () => async ({ test }) => ({
        test: await mockUpdate(test)
      }),
      mock: (test: string) => () => {
        mockAction(test);

        return {};
      },
      error: () => () => 'error' as any
    }),
    mockUpdate,
    mockAction
  };
};

it('should initialize the dakpan', () => {
  const { Provider, Consumer, actions } = createMockDakpan();

  expect(Provider).toBeDefined();
  expect(Consumer).toBeDefined();
  expect(actions).toBeDefined();
  expect(actions.append).toBeDefined();
});

it('should throw when an action is dispatched before its provider is mounted', () => {
  const { actions } = createMockDakpan();

  expect(actions.append.e('!')).toThrow('You may not dispatch an action before its provider is mounted');
});

it('should update the store', () => {
  const { Provider, Consumer, actions } = createMockDakpan();

  const wrapper = mount(
    <Provider>
      <Consumer>
        {({ hello }) => hello}
      </Consumer>
    </Provider>
  );

  expect(wrapper).toMatchSnapshot();

  actions.append('!');

  expect(wrapper.update()).toMatchSnapshot();
});

it('should update the store asynchronously', async () => {
  const { Provider, Consumer, actions, mockUpdate } = createMockDakpan();

  const wrapper = mount(
    <Provider>
      <Consumer>
        {({ test }) => test}
      </Consumer>
    </Provider>
  );

  expect(wrapper).toMatchSnapshot();

  await actions.increment();

  expect(mockUpdate).toBeCalledWith(0);
  expect(wrapper.update()).toMatchSnapshot();
});

it('should update the store as event handler', () => {
  const { Provider, Consumer, actions } = createMockDakpan();

  const wrapper = mount(
    <Provider>
      <Consumer>
        {({ hello }) => (
          <>
            <span>{hello}</span>
            <button onClick={actions.append.e('!')}/>
          </>
        )}
      </Consumer>
    </Provider>
  );

  expect(wrapper).toMatchSnapshot();
  wrapper.find('button').simulate('click');
  expect(wrapper).toMatchSnapshot();
});

it('should call an action with the correct parameters', () => {
  const { Provider, actions, mockAction } = createMockDakpan();

  mount(
    <Provider>
      <div/>
    </Provider>
  );

  actions.mock('hello!');
  const e = actions.mock.e('hello!!');

  expect(mockAction).toBeCalledWith('hello!');

  e();

  expect(mockAction).toBeCalledWith('hello!!');
});

it('should throw when the action doesn\'t return an object', () => {
  const { Provider, actions } = createMockDakpan();

  mount(
    <Provider>
      <div/>
    </Provider>
  );

  expect(actions.error.e()).toThrow('Actions may only return objects');
});

it('should pass the actions to the consumer', () => {
  const { Provider, Consumer } = createMockDakpan();

  const wrapper = mount(
    <Provider>
      <Consumer>
        {({ hello }, { append }) => (
          <>
            <span>{hello}</span>
            <button onClick={append.e('!')}/>
          </>
        )}
      </Consumer>
    </Provider>
  );

  expect(wrapper).toMatchSnapshot();
  wrapper.find('button').simulate('click');
  expect(wrapper).toMatchSnapshot();
});

it('should return the updated state', async () => {
  const { Provider, actions } = createMockDakpan();

  mount(
    <Provider>
      <div/>
    </Provider>
  );

  expect(actions.append('!')).toMatchSnapshot();
  expect(await actions.increment()).toMatchSnapshot();
});

it('should pass the state to a stateless component using a hoc', () => {
  const { Provider, withDakpan } = createMockDakpan();

  type Test = {
    test: string
  };

  const Component = withDakpan(({ hello }, { append }) => ({
    hello,
    append
  }))<Test>(({ test, hello, append }) => (
    <>
      <span>{test}</span>
      <span>{hello}</span>
      <button onClick={append.e('!')}/>
    </>
  ));

  const wrapper = mount(
    <Provider>
      <Component test="prop"/>
    </Provider>
  );

  expect(wrapper).toMatchSnapshot();
});

it('should pass the state to a stateful component using a hoc', () => {
  const { Provider, actions, withDakpan } = createMockDakpan();

  type Props = {
    test: string,
    hello: string,
    append: typeof actions.append
  };

  const TestComponent = withDakpan(({ hello }, { append }) => ({
    hello,
    append
  }))(class extends Component<Props> {
    public componentDidMount() {
      const { append } = this.props;

      append('!');
    }

    public render() {
      const { test, hello } = this.props;

      return (
        <>
          <span>{test}</span>
          <span>{hello}</span>
        </>
      );
    }
  });

  const wrapper = mount(
    <Provider>
      <TestComponent test="prop"/>
    </Provider>
  );

  expect(wrapper).toMatchSnapshot();
});
