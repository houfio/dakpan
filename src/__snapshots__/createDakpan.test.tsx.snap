// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should pass the actions to the consumer 1`] = `
<Component>
  <Component>
    <span>
      world
    </span>
    <button
      onClick={[Function]}
    />
  </Component>
</Component>
`;

exports[`should pass the actions to the consumer 2`] = `
<Component>
  <Component>
    <span>
      world!
    </span>
    <button
      onClick={[Function]}
    />
  </Component>
</Component>
`;

exports[`should pass the state to a stateful component using a hoc 1`] = `
<Component>
  <Component
    test="prop"
  >
    <Component>
      <Component
        append={[Function]}
        hello="world!"
        test="prop"
      >
        <span>
          prop
        </span>
        <span>
          world!
        </span>
      </Component>
    </Component>
  </Component>
</Component>
`;

exports[`should pass the state to a stateless component using a hoc 1`] = `
<Component>
  <Component
    test="prop"
  >
    <Component>
      <Component
        append={[Function]}
        hello="world"
        test="prop"
      >
        <span>
          prop
        </span>
        <span>
          world
        </span>
        <button
          onClick={[Function]}
        />
      </Component>
    </Component>
  </Component>
</Component>
`;

exports[`should return the updated state 1`] = `
Object {
  "hello": "world!",
  "test": 0,
}
`;

exports[`should return the updated state 2`] = `
Object {
  "hello": "world!",
  "test": 1,
}
`;

exports[`should update the store 1`] = `
<Component>
  <Component>
    world
  </Component>
</Component>
`;

exports[`should update the store 2`] = `
<Component>
  <Component>
    world!
  </Component>
</Component>
`;

exports[`should update the store as event handler 1`] = `
<Component>
  <Component>
    <span>
      world
    </span>
    <button
      onClick={[Function]}
    />
  </Component>
</Component>
`;

exports[`should update the store as event handler 2`] = `
<Component>
  <Component>
    <span>
      world!
    </span>
    <button
      onClick={[Function]}
    />
  </Component>
</Component>
`;

exports[`should update the store asynchronously 1`] = `
<Component>
  <Component>
    0
  </Component>
</Component>
`;

exports[`should update the store asynchronously 2`] = `
<Component>
  <Component>
    1
  </Component>
</Component>
`;
